feat: Add regions to `Sender` and `Receiver`

paulbutcher opened the pull request:
This is the start of an attempt to implement the "Modify Sender[t] and Receiver[t] types to carry a region (e.g. Sender[t, r])." task of #4850, but definitely not there yet.

I'm aware that passing `Type.False` for the region parameter to `Type.mkSender` and `Type.mkReceiver` isn't the right thing to do, but I don't see what my options are here without adding a region parameter to `Channel.buffered` etc. because I don't have a region type variable otherwise?

I suspect I'm missing something: could someone point me in the right direction?

magnus-madsen commented:
Here: 

```
pub def buffered(n: Int32): (Sender[t, r], Receiver[t, r]
```

you need an extra argument `r: Region[r]`. 

Then you should modify `CHANNEL_NEW` to take two arguments.

magnus-madsen commented:
Otherwise your changes look to be going in the right direction.

paulbutcher commented:
> you need an extra argument r: Region[r].

That is obviously the right long-term solution, but I was trying to do what you suggested in #4850 and first have a PR which adds the region to `Sender` and `Receiver` only without modifying the signature of the channel functions. I assumed that that there was some way to do that, given that that's what you requested.

So I'll change this to do both of the first two tasks within #4850. But be aware that this is going to turn into a monster PR because it will entail changing every use of channels. And I know how much you like monster PRs 😜

magnus-madsen commented:
> That is obviously the right long-term solution, but I was trying to do what you suggested in https://github.com/flix/flix/issues/4850 and first have a PR which adds the region to Sender and Receiver only without modifying the signature of the channel functions. I assumed that that there was some way to do that, given that that's what you requested.

Ah yes. Uhm. Let me think.

magnus-madsen commented:
Can you do this instead (to avoid the monster PR):

Have buffered/unbuffered return with `Static` as the region?

paulbutcher commented:
I'll see if I can make that work 👍

paulbutcher commented:
I think that this is now ready for review.

paulbutcher commented on main/src/ca/uwaterloo/flix/language/phase/Lowering.scala:
```diff
@@ -1361,7 +1361,7 @@ object Lowering {
         val pat = mkTuplePattern(List(LoweredAst.Pattern.Cst(Ast.Constant.Int32(i), Type.Int32, loc), LoweredAst.Pattern.Var(locksSym, locksType, loc)), loc)
         val getTpe = Type.eraseTopAliases(chan.tpe) match {
-          case Type.Apply(_, t, _) => t
+          case Type.Apply(Type.Apply(_, t, _), _, _) => t
```
@magnus-madsen I'm not really happy with this: it feels rather fragile to me. But I'm not sure that there's any better way? Suggestions welcome.

magnus-madsen replied:
This is acceptable. Your other alternative is `.typeArguments.head`, but what you have is fine.

magnus-madsen commented on main/src/library/Concurrent/Channel.flix:
```diff
@@ -63,14 +63,14 @@ namespace Concurrent/Channel {
     ///
     @Internal
-    pub enum Mpmc[a](
+    pub enum Mpmc[a: Type, _: Region](
```
I think you have two options here:

- Either you just drop the region entirely (and ignore it) or
- You introduce the region, but then you should also use it, i.e. in the definition of MutDeque.

magnus-madsen replied:
If you keep the region parameter then you should use the actual region available in Lowering.

magnus-madsen replied:
The right thing would be to use the region, but the easy thing (which I lean towards), is to ignore it.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Lowering.scala:
```diff
@@ -598,7 +598,7 @@ object Lowering {
       val e = visitExp(exp)
       val t = visitType(tpe)
-      val chTpe = mkChannelTpe(elmTpe, loc)
+      val chTpe = mkChannelTpe(elmTpe, Type.False, loc)
```
See comment below about Mpmc.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Typer.scala:
```diff
@@ -1604,14 +1604,14 @@ object Typer {
           (constrs, tpe, _, eff) <- visitExp(exp)
           _ <- expectTypeM(expected = Type.Int32, actual = tpe, exp.loc)
-          resultTyp <- liftM(Type.mkTuple(List(Type.mkSender(elmType, loc), Type.mkReceiver(elmType, loc)), loc))
+          resultTyp <- liftM(Type.mkTuple(List(Type.mkSender(elmType, Type.False, loc), Type.mkReceiver(elmType, Type.False, loc)), loc))
```
Here you should not use Type.False but create a fresh type variable and unify with it.

paulbutcher replied:
For this and the following similar comments, I see why this would be the right thing to do, but am I not going to then have to immediately discard that code as soon as I add a region parameter to the channel functions (because I'll be unifying against that parameter and won't have to create one here?)

Or am I misunderstanding?

magnus-madsen replied:
It would be the opposite. (The change I propose would be forwards compatible).

The point is that only GetChannel takes a region parameter. For the other channel operations the region is inferred by the type system (hence the need for type variables that get unified.)

paulbutcher replied:
Oh! Then I am definitely misunderstanding. 

Do you mean GetChannel? Or NewChannel? Above you were talking about adding a `r: Region[r]` parameter to `Channel.buffered`?

Forgive me, I'm confused.

magnus-madsen replied:
Maybe I am confused. My point was for everything but NEW_CHANNEL we should infer the region (i.e. not use Static/False, but rather a type variable).

magnus-madsen replied:
Here I am only talking about what occurs in Typer and only the AST nodes.

paulbutcher replied:
Ah, NewChannel. That makes much more sense. OK, I think I get it, thanks.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Typer.scala:
```diff
@@ -1604,14 +1604,14 @@ object Typer {
       case KindedAst.Expression.GetChannel(exp, tvar, loc) =>
         val elmVar = Type.freshVar(Kind.Star, loc, text = FallbackText("elm"))
-        val channelType = Type.mkReceiver(elmVar, loc)
+        val channelType = Type.mkReceiver(elmVar, Type.False, loc)
```
Here also you should have a regVar next to elmVar.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Typer.scala:
```diff
@@ -1623,7 +1623,7 @@ object Typer {
       case KindedAst.Expression.PutChannel(exp1, exp2, loc) =>
         val elmVar = Type.freshVar(Kind.Star, loc, text = FallbackText("elm"))
-        val channelType = Type.mkSender(elmVar, loc)
+        val channelType = Type.mkSender(elmVar, Type.False, loc)
```
And regVar next to elmVar.

paulbutcher commented:
I think that this might be there now @magnus-madsen

magnus-madsen commented on main/src/library/Concurrent/Channel.flix:
```diff
@@ -63,7 +63,7 @@ namespace Concurrent/Channel {
     ///
     @Internal
-    pub enum Mpmc[a](
+    pub enum Mpmc[a: Type](
```
This a: Type is no longer needed.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Typer.scala:
```diff
@@ -1601,17 +1601,20 @@ object Typer {
 
 
       case KindedAst.Expression.NewChannel(exp, elmType, loc) =>
+        val regionVar = Type.freshVar(Kind.Bool, loc, text = FallbackText("region"))
```
This should be False.

paulbutcher replied:
Sorry Magnus, what's the "this" that should be False? Are you suggesting that I shouldn't have a `regionVar` here at all and instead pass `Type.False` to `mkSender` and `mkReceiver` below. That contradicts what you said earlier, so I guess not? But then I don't know what else you might mean?

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Typer.scala:
```diff
@@ -1601,17 +1601,20 @@ object Typer {
         } yield (constrs, resultTyp, resultPur, resultEff)
 
       case KindedAst.Expression.GetChannel(exp, tvar, loc) =>
+        val regionVar = Type.freshVar(Kind.Bool, loc, text = FallbackText("region"))
```
This is correct.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Typer.scala:
```diff
@@ -1622,8 +1625,9 @@ object Typer {
         } yield (constrs, resultTyp, resultPur, resultEff)
 
       case KindedAst.Expression.PutChannel(exp1, exp2, loc) =>
+        val regionVar = Type.freshVar(Kind.Bool, loc, text = FallbackText("region"))
```
This is correct.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Typer.scala:
```diff
@@ -1645,7 +1649,7 @@ object Typer {
               (chanConstrs, chanType, _, chanEff) <- visitExp(chan)
               (bodyConstrs, bodyType, _, bodyEff) <- visitExp(body)
-              _ <- unifyTypeM(chanType, Type.mkReceiver(sym.tvar, sym.loc), sym.loc)
+              _ <- unifyTypeM(chanType, Type.mkReceiver(sym.tvar, Type.False, sym.loc), sym.loc)
```
This should be regionVar

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Typer.scala:
```diff
@@ -2255,7 +2259,7 @@ object Typer {
         val pur = Type.Impure
         val eff = e.eff
-        TypedAst.Expression.NewChannel(e, Type.mkTuple(List(Type.mkSender(elmTpe, loc), Type.mkReceiver(elmTpe, loc)), loc), elmTpe, pur, eff, loc)
+        TypedAst.Expression.NewChannel(e, Type.mkTuple(List(Type.mkSender(elmTpe, Type.False, loc), Type.mkReceiver(elmTpe, Type.False, loc)), loc), elmTpe, pur, eff, loc)
```
I think instead of all this, we should equip NewChannel with a tvar, and the unify with that, and simply use subst(tvar) here.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Typer.scala:
```diff
@@ -1601,17 +1601,20 @@ object Typer {
           (constrs, tpe, _, eff) <- visitExp(exp)
           _ <- expectTypeM(expected = Type.Int32, actual = tpe, exp.loc)
-          resultTyp <- liftM(Type.mkTuple(List(Type.mkSender(elmType, loc), Type.mkReceiver(elmType, loc)), loc))
+          resultTyp <- liftM(Type.mkTuple(List(Type.mkSender(elmType, regionVar, loc), Type.mkReceiver(elmType, regionVar, loc)), loc))
```
Here we are allocating a new channel (and for now) that is always in the static region, hence regionVar here should really be Type.False. (This will change when you later add regions to NEW_CHANNEL)

paulbutcher commented:
Oh! I wasn't expect you to merge this yet 🤣. I haven't addressed your comment above: "I think instead of all this, we should equip NewChannel with a tvar, and the unify with that, and simply use subst(tvar) here."

But... I was going to say that I'm not sure that I see how this makes things better. We are already inferring `elmType`. If we infer the tuple type instead (which I think is what you're getting at) it just moves the complexity elsewhere (because we need to pick the element type out of the tuple type).

Or am I missing something?

magnus-madsen commented:
> Oh! I wasn't expect you to merge this yet 🤣. I haven't addressed your comment above: "I think instead of all this, we should equip NewChannel with a tvar, and the unify with that, and simply use subst(tvar) here."
> 
> But... I was going to say that I'm not sure that I see how this makes things better. We are already inferring `elmType`. If we infer the tuple type instead (which I think is what you're getting at) it just moves the complexity elsewhere (because we need to pick the element type out of the tuple type).
> 
> Or am I missing something?

I think this will become clearer later once we care about both the elmtype and region. Then we can revisit.
